src/App/Application.cpp

```diff
@@ -2217,6 +2219,8 @@ void Application::initTypes()
     new Base::ExceptionProducer<Base::CADKernelError>;
     new Base::ExceptionProducer<Base::RestoreError>;
     new Base::ExceptionProducer<Base::PropertyError>;
+
+    ServiceProvider::get().implement<CenterOfMassProvider>(new NullCenterOfMass);

```

Comment:
ServiceProvider is not really singleton - you can instantiate multiple instances of it if it is needed for any reason (for example to provide scoped provider for typical service locator use). There is however one global service provider implemented as singleton that should be used to acquire global services. Ideally it should probably be part of `Application` class.

I also don't think think that namespace here is necessary - I consider ServiceProvider to be implementation detail. The goal of this function is to obtain instance of service implementing given interface - Service Provider class is only means to do so. If anything, I'd call the namespace simply `Service`. However, given that the name is already quite lengthy I would not want to make it any longer - ideally it should fit in one line.

As for the names:
 - I'll surely rename `Base::implementContract` into `Base::implementService` or something like thatas I fortgot to rename that method originally. I don't like `addService` as it does not add service, only an implementation of service - service is defined by the abstract interface class.
 - `Base::provideImplementation()` - I am open to change that name, `provide` seems like good word, but `acquireService` also seems nice and is a bit shorter.
 - I don't like `acquireAllServices` on the same reasoning as first point. This PR introduces 2 services: `CenterOfMassProvider` and `SubObjectPlacementProvider` with 2 and 1 implementation respectively. For method named `acquireAllServices` I'd assume to obtain both `CenterOfMassProvider` and `SubObjectPlacementProvider` not all implementations of lets say `CenterOfMassProvider`

Reply:
Ok, I can see that the service provider is not really a singleton. Indeed you can construct one yourself. However, by providing a static get function, you really communicate to readers of the code that the class is meant to be a singleton. The public constructor communicates the opposite, so for me it feels like mixed signals here and clearly I overlooked the public constructor and assumed it was a singleton based on the static `get()`. If there is one service provider meant to be global, but you can also construct service providers for limited scope, I would prefer to simply have a global instance available. I don't mind global variables for things that are meant to be global.

I guess I wasn't clear on that but you don't address why there are two ways of accessing the global instance. One way is what you do above:
- `ServiceProvider::get().implement<CenterOfMassProvider>(new NullCenterOfMass);`
- `Base::implementContract<CenterOfMassProvider>(new NullCenterOfMass);`

I find the second one more logical for this case (it is accessing the global one). 

About the namespace, I don't find it necessary if the ServiceProvider keeps on being a class.

Regarding naming. I can see your point, but `Base::implementService` sounds to me as if this function is going to create an implementation for a service which is not the case. I can see the difference between a service and an implementation of the service, but from the viewpoint of the user of the function, you created an implementation of a service and you're adding it to the service provider. At the moment you created the implementation of the service, the implementation will act as a service, so when the service switches from being an implementation to when it becomes a service (in any case before calling `acquireService()`) is splitting hairs a bit. Perhaps the most accurate would be `addServiceImplementation()`, but that is a bit long. I would simply call it `addService()`, but it is not critical to me.

Regarding your comment on `acquireAllServices()`: I don't think that is the case, because we still have the template parameter that communicates your acquiring all services of `CenterOfMassProvider` or `SubObjectPlacementProvider`.